### ISSUE TITLE
Protocol configurations

### ISSUE THREAD

**Issue by mortendahl** [mortendahl/tf-encrypted#239]
We are starting to have several parameters and protocols to specify; how do we deal with configuration of these?

**Comment by mortendahl** [comment]
One thing to keep in mind is that the graph is constructed based on these properties, ie. choice of protocol, choice of tensor, etc.

**Comment by ianlivingstone** [comment]
> One thing to keep in mind is that the graph is constructed based on these properties, ie. choice of protocol, choice of tensor, etc.

Good point - so its really graph creation configuration then which happens beforehand. 

What is the ideal world here for a user? How much should they have to change to switchbacking tensor types? Is it a matter of configuration or should the fact they're using a crt or int64 tensor bubble all the way to the top into their code or should it be a runtime configuration concern?

**Comment by bendecoste** [comment]
I think sensible defaults is probably the solution here.  It is likely that int64 will always be the default for both protocols.  However you may wish to require int100 in an odd ring.  Are there any other types right now that a user might want? perhaps int32 in some cases?

There are also other tensor types that are going to be even weirder for users and we may wish to hide: PrimeTensor, OddImplicit, et

**Comment by mortendahl** [comment]
> Good point - so its really graph creation configuration then which happens beforehand.

Exactly! You can change it by switching to another protocol instance during graph construction though.

> What is the ideal world here for a user? How much should they have to change to switchbacking tensor types? Is it a matter of configuration or should the fact they're using a crt or int64 tensor bubble all the way to the top into their code or should it be a runtime configuration concern?

Agree that sensible defaults here would be good (and int64 is not a bad choice).

I was playing a bit with the idea of seeing the protocols as just another form of what we now call tensor factories, ie the are not something the user would typically instantiate on his own but simple just singletons (advanced users might want to instantiate their own). This way, a Pond tensor and a SecureNN tensor is just differentiated by a dtype (one pointing to a Pond instance, one to a SecureNN instance). One question here is how conversion of tensors from the different protocols should work (this is quite common, say converting from SPDZ to GC)
